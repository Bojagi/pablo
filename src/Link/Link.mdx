---
name: Link
route: /link
menu: Components
---

import { Props, Playground } from 'docz'
import { Box } from '../Box/Box.tsx'
import { Link } from './Link.tsx'

# Image

The `Link` component is a styled version of the normal HTML anchor element. It can be easily used with the 
`Link` component from `react-router`.

## Examples

### Plain link

The `Link` component supports all props a normal `<a>` component would do.

<Playground>
  <Link href="https://bojagi.io" target="_blank">Go to Bojagi</Link>
</Playground>

### Using it together with react-router

You can use margin to give images a bit of space around them

```jsx
import { Link } from '@bojagi/pablo/Link';
// This can also be the Link from gatsby or next
import { Link as RouterLink } from 'react-router-dom';

const InternalLink = props => <Link {...props} as={RouterLink} />
```

## Import

```js
import { Link } from '@bojagi/pablo/Link';
```

## Props

<Props of={Link} />
