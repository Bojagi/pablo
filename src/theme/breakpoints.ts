export type Breakpoint = 'sm' | 'md' | 'lg' | 'xl';

export const breakpointNames: Breakpoint[] = ['sm', 'md', 'lg', 'xl'];

export type BreakpointsArray = [number, number, number, number];

export type Breakpoints = BreakpointsArray &
  Record<Breakpoint, number> & {
    breakpointNames: Breakpoint[];
  };

export function createBreakpoints(
  array: BreakpointsArray,
  usedBreakpointNames: Breakpoint[] = breakpointNames
): Breakpoints {
  const newBreakpoints = ([...array] as unknown) as Breakpoints;
  usedBreakpointNames.forEach((bp, index) => {
    newBreakpoints[bp] = newBreakpoints[index];
  });
  newBreakpoints.breakpointNames = usedBreakpointNames;

  return newBreakpoints as Breakpoints;
}

export const defaultBreakpoints: BreakpointsArray = [600, 1000, 1200, 1920];

export const breakpoints = createBreakpoints(defaultBreakpoints, breakpointNames);
