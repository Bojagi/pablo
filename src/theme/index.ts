export * from './defaultTheme';
export * from './context';
export * from './PabloThemeProvider';
export * from './useComponentStyle';
