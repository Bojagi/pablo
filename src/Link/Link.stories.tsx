import React from 'react';
import { Subtitle } from '../Typography';
import { Link } from './Link';

export default {
  title: 'Link',
};

export const SimpleLink = () => (
  <Subtitle>
    <Link href="https://bojagi.io" target="_blank">
      Go to Bojagi
    </Link>
  </Subtitle>
);
