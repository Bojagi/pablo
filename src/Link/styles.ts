import { getColor } from '../styleHelpers/getColor';
import { Style } from '../theme/types';

export type FontStyleValues =
  | 'normal'
  | 'italic'
  | 'oblique'
  | `oblique ${number}deg`
  | 'inherit'
  | 'initial'
  | 'unset';

export interface BaseLinkStyle {
  color: Style;
  fontStyle: FontStyleValues;
  textDecoration: string;
  fontWeight: string | number;
}

export interface LinkStyles extends BaseLinkStyle {
  hover: BaseLinkStyle;
  focus: BaseLinkStyle;
  visited: BaseLinkStyle;
}

const BASE_LINK_STYLES: BaseLinkStyle = {
  color: getColor('brand', 'main'),
  textDecoration: 'none',
  fontStyle: 'normal',
  fontWeight: 'inherit',
};

export const linkStyles: LinkStyles = {
  ...BASE_LINK_STYLES,
  hover: {
    ...BASE_LINK_STYLES,
    textDecoration: 'underline',
  },
  visited: {
    ...BASE_LINK_STYLES,
  },
  focus: {
    ...BASE_LINK_STYLES,
    textDecoration: 'underline',
  },
};
