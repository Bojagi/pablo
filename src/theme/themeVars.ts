import isObject from 'isobject';
import { breakpointVars } from './breakpoints';
import { colorVars } from './colors';
import { typographyVars } from './typography';

export const themeVarNames = {
  typography: typographyVars,
  colors: colorVars,
  breakpoints: breakpointVars,
};

function createThemeReferences<T extends Record<string, any>>(obj: T): T {
  return Object.entries(obj).reduce(
    (acc, [key, value]) => ({
      ...acc,
      [key]: isObject(value) ? createThemeReferences(value as T) : `var(--${value})`,
    }),
    {} as any
  );
}

export const themeVars = createThemeReferences(themeVarNames);
