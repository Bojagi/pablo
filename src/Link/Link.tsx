import styled from '@emotion/styled';
import { baseStyle } from '../shared/baseStyle';
import { getComponentStyle } from '../styleHelpers';
import { BaseProps } from '../types';
import { getCustomStyles } from '../utils/useCustomStyles';
import { LinkStyleProperties } from './styles';

export type LinkProps = BaseProps<LinkStyleProperties>;

export const Link = styled.a<LinkProps>`
  ${baseStyle}
  color: ${getComponentStyle(['link', 'color'])};
  text-decoration: ${getComponentStyle(['link', 'textDecoration'])};
  font-style: ${getComponentStyle(['link', 'fontStyle'])};
  font-weight: ${getComponentStyle(['link', 'fontWeight'])};
  outline: 0;
  ${getCustomStyles('link.styles', 'root')}

  &:hover {
    color: ${getComponentStyle(['link', 'hover', 'color'])};
    text-decoration: ${getComponentStyle(['link', 'hover', 'textDecoration'])};
    font-style: ${getComponentStyle(['link', 'hover', 'fontStyle'])};
    font-weight: ${getComponentStyle(['link', 'hover', 'fontWeight'])};
    ${getCustomStyles('link.styles', 'hover')}
  }
  &:visited {
    color: ${getComponentStyle(['link', 'visited', 'color'])};
    text-decoration: ${getComponentStyle(['link', 'visited', 'textDecoration'])};
    font-style: ${getComponentStyle(['link', 'visited', 'fontStyle'])};
    font-weight: ${getComponentStyle(['link', 'visited', 'fontWeight'])};
    ${getCustomStyles('link.styles', 'visited')}
  }
  &:focus {
    color: ${getComponentStyle(['link', 'focus', 'color'])};
    text-decoration: ${getComponentStyle(['link', 'focus', 'textDecoration'])};
    font-style: ${getComponentStyle(['link', 'focus', 'fontStyle'])};
    font-weight: ${getComponentStyle(['link', 'focus', 'fontWeight'])};
    ${getCustomStyles('link.styles', 'focus')}
  }
`;
