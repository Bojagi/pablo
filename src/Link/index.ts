import './styles';

export * from './Link';
