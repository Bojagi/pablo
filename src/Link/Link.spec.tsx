import { render } from '@testing-library/react';
import React from 'react';
import { PabloThemeProvider } from '../theme';
import { themeVars } from '../theme/themeVars';
import { Link } from './Link';

test('Render Link with the right styles', () => {
  const { container } = renderComponent({});
  const linkElem = container.firstChild;
  expect(linkElem).toMatchSnapshot();

  // Normal state
  expect(linkElem).toHaveStyleRule('color', themeVars.colors.brand.main);
  expect(linkElem).toHaveStyleRule('text-decoration', 'none');
  expect(linkElem).toHaveStyleRule('font-style', 'normal');
  expect(linkElem).toHaveStyleRule('font-weight', 'inherit');

  // Hover state
  expect(linkElem).toHaveStyleRule('color', themeVars.colors.brand.main, {
    modifier: ':hover',
  });
  expect(linkElem).toHaveStyleRule('text-decoration', 'underline', {
    modifier: ':hover',
  });
  expect(linkElem).toHaveStyleRule('font-style', 'normal', {
    modifier: ':hover',
  });
  expect(linkElem).toHaveStyleRule('font-weight', 'inherit', {
    modifier: ':hover',
  });

  // Focus state
  expect(linkElem).toHaveStyleRule('color', themeVars.colors.brand.main, {
    modifier: ':focus',
  });
  expect(linkElem).toHaveStyleRule('text-decoration', 'underline', {
    modifier: ':focus',
  });
  expect(linkElem).toHaveStyleRule('font-style', 'normal', {
    modifier: ':focus',
  });
  expect(linkElem).toHaveStyleRule('font-weight', 'inherit', {
    modifier: ':focus',
  });

  // Visited state
  expect(linkElem).toHaveStyleRule('color', themeVars.colors.brand.main, {
    modifier: ':visited',
  });
  expect(linkElem).toHaveStyleRule('text-decoration', 'none', {
    modifier: ':visited',
  });
  expect(linkElem).toHaveStyleRule('font-style', 'normal', {
    modifier: ':visited',
  });
  expect(linkElem).toHaveStyleRule('font-weight', 'inherit', {
    modifier: ':visited',
  });
});

function renderComponent(props) {
  return render(
    <PabloThemeProvider>
      <Link href="https://hello.goodbye" {...props}>
        Hello
      </Link>
    </PabloThemeProvider>
  );
}
