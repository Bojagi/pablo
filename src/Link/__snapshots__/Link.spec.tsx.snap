// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Link with the right styles 1`] = `
.c0 {
  box-sizing: border-box;
  color: var(--pbl-theme-colors-brand-main);
  -webkit-text-decoration: none;
  text-decoration: none;
  font-style: normal;
  font-weight: inherit;
  outline: 0;
}

.c0:hover {
  color: var(--pbl-theme-colors-brand-main);
  -webkit-text-decoration: underline;
  text-decoration: underline;
  font-style: normal;
  font-weight: inherit;
}

.c0:visited {
  color: var(--pbl-theme-colors-brand-main);
  -webkit-text-decoration: none;
  text-decoration: none;
  font-style: normal;
  font-weight: inherit;
}

.c0:focus {
  color: var(--pbl-theme-colors-brand-main);
  -webkit-text-decoration: underline;
  text-decoration: underline;
  font-style: normal;
  font-weight: inherit;
}

<a
  class="c0"
  href="https://hello.goodbye"
>
  Hello
</a>
`;
